multicluster/cmd/service-mirror/main.go

```diff
@@ -84,79 +103,159 @@ func Main(args []string) {
+								if err != nil {
+									log.Errorf("Failed to load remote cluster credentials: %s", err)
+								}
+								err = restartClusterWatcher(ctx, link, *namespace, creds, controllerK8sAPI, *requeueLimit, *repairPeriod, metrics, *enableHeadlessSvc)
+								if err != nil {
+									// failed to restart cluster watcher; give a bit of slack
+									// and restart the link watch to give it another try
+									log.Error(err)
+									time.Sleep(linkWatchRestartAfter)
+									linkWatch.Stop()
+								}
+							case watch.Deleted:
+								log.Infof("Link %s deleted", linkName)
+								cleanupWorkers()
+							default:
+								log.Infof("Ignoring event type %s", event.Type)
 							}
-						default:
-							log.Infof("Ignoring event type %s", event.Type)
 						}
+					default:
+						log.Errorf("Unknown object type detected: %+v", obj)
 					}
-				default:
-					log.Errorf("Unknown object type detected: %+v", obj)
 				}
 			}
 		}
 	}
+
+	hostname, found := os.LookupEnv("HOSTNAME")
+	if !found {
+		log.Fatal("Failed to fetch 'HOSTNAME' environment variable")
+	}
+
+	lock := &resourcelock.LeaseLock{
+		LeaseMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("service-mirror-write-%s", linkName),
+			Namespace: *namespace,
+			Labels: map[string]string{
+				"component":                      "linkerd-service-mirror",
+				"mirror.linkerd.io/cluster-name": linkName,
+			},
+		},
+		Client: k8sAPI.CoordinationV1(),
+		LockConfig: resourcelock.ResourceLockConfig{
+			Identity: hostname,
+		},
+	}
+
+	for loop := true; loop; {
+		// RunOrDie will block until a lease is acquired. Once a lease is
+		// acquired, the main watcher loop will start a Link watch. If the lease
+		// is lost, clean-up watchers and re-attempt to acquire.
+		leaderelection.RunOrDie(rootCtx, leaderelection.LeaderElectionConfig{
+			// When runtime context is cancelled, lock will be released. Implies any
+			// code guarded by the least _must_ finish before cancelling.
+			ReleaseOnCancel: true,
+			Lock:            lock,
+			LeaseDuration:   LEASE_DURATION,
+			RenewDeadline:   LEASE_RENEW_DEADLINE,
+			RetryPeriod:     LEASE_RETRY_PERIOD,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(ctx context.Context) {
+					// Rely on cancellation to be propagated from leader elector
+					// in the worker
+					log.Info("Starting controller loop")
+					run(ctx)
+				},
+				OnStoppedLeading: func() {
+					log.Infof("%s released lease", hostname)
+				},
+				OnNewLeader: func(identity string) {
+					if identity == hostname {
+						log.Infof("%s acquired lease", hostname)
+					}
+				},
+			},
+		})
+
+		select {
+		// If we have just lost the lease, and we have received a shutdown
+		// signal, break the loop and gracefully exit. We can guarantee at this
+		// point resources have been released.
+		case <-rootCtx.Done():
+			loop = false
+		default:
+			// Do nothing once a lease is lost, keeping spinning and don't block
```

Comment:
This comment is a bit misleading.  When we lose the lease, we loop and attempt to acquire it again.